pkg/innerring/nns.go

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+var errDomainNotFound = errors.New("domain not found")
+
+type errWrongStackItemType struct {
+	expected, got stackitem.Type
+}
+
+func wrongStackItemTypeError(expected, got stackitem.Type) errWrongStackItemType {
+	return errWrongStackItemType{
+		expected: expected,
+		got:      got,
+	}
+}
+
+func (x errWrongStackItemType) Error() string {
+	return fmt.Sprintf("wrong type: expected %s, got %s", x.expected, x.got)
+}
+
+type errInvalidNumberOfStructFields struct {
+	expected, got int
+}
+
+func invalidNumberOfStructFieldsError(expected, got int) errInvalidNumberOfStructFields {
+	return errInvalidNumberOfStructFields{
+		expected: expected,
+		got:      got,
+	}
+}
+
+func (x errInvalidNumberOfStructFields) Error() string {
+	return fmt.Sprintf("invalid number of struct fields: expected %d, got %d", x.expected, x.got)
+}
+
+type errInvalidStructField struct {
+	index uint
+	cause error
+}
+
+func invalidStructFieldError(index uint, cause error) errInvalidStructField {
+	return errInvalidStructField{
+		index: index,
+		cause: cause,
+	}
+}
+
+func (x errInvalidStructField) Error() string {
+	return fmt.Sprintf("invalid struct field #%d: %v", x.index, x.cause)
+}
+
+func (x errInvalidStructField) Unwrap() error {
+	return x.cause
+}
+
+type errWrongNNSRecordType struct {
+	expected, got *big.Int
+}
+
+func (x errWrongNNSRecordType) Error() string {
+	return fmt.Sprintf("wrong type of record: expected %v, got %v", x.expected, x.got)
+}
+
+func wrongNNSRecordTypeError(expected, got *big.Int) errWrongNNSRecordType {
+	return errWrongNNSRecordType{
+		expected: expected,
+		got:      got,
+	}
+}
+
+type errInvalidNNSDomainRecord struct {
+	domain string
+	record string
+	cause  error
+}
+
+func invalidNNSDomainRecordError(domain, record string, cause error) errInvalidNNSDomainRecord {
+	return errInvalidNNSDomainRecord{
+		domain: domain,
+		record: record,
+		cause:  cause,
+	}
+}
+
+func (x errInvalidNNSDomainRecord) Error() string {
+	if x.record != "" {
+		return fmt.Sprintf("invalid record %q of the NNS domain %q: %v", x.record, x.domain, x.cause)
+	}
+
+	return fmt.Sprintf("invalid record of the NNS domain %q: %v", x.domain, x.cause)
+}
+
+func (x errInvalidNNSDomainRecord) Unwrap() error {
+	return x.cause
+}
+
+// CheckDomainRecord calls iterating 'getAllRecords' method of the parameterized
+// Neo smart contract passing the given domain name. If contract throws 'token
+// not found' exception, CheckDomainRecord returns errDomainNotFound. Each value
+// in the resulting iterator is expected to be structure with at least 3 fields.
+// If any value has the 2nd field is a number equal to 16 (TXT record type in
+// the NNS) and the 3rd one is a string equal to the specified record,
+// CheckDomainRecord returns nil. Otherwise,
+// [privatedomains.ErrMissingDomainRecord] is returned.
+func (x *neoFSNNS) CheckDomainRecord(domain string, record string) error {
+	sessionID, iter, err := x.contract.GetAllRecords(domain)
```

Comment:
This doesn't justify all of this code. Resolve can have iterators as well, but it might not have them too,  https://github.com/nspcc-dev/neofs-contract/pull/266/commits/b1f5864942138d8667badd0f7c583417a5bc593c

Reply:
done